perl/cpan/private/carton.bzl

```
@@ -0,0 +1,199 @@
+    visibility = ["//visibility:public"],
+)
+"""
+
+def _perl_cpan_archive_impl(repository_ctx):
+    results = repository_ctx.download_and_extract(
+        repository_ctx.attr.urls,
+        sha256 = repository_ctx.attr.sha256,
+        stripPrefix = repository_ctx.attr.strip_prefix,
+    )
+
+    output = ""
+    includes = []
+    lib_path = repository_ctx.path("lib")
+    if lib_path.exists:
+        includes = ["lib"]
+    else:
+        # In the event a lib directory doesn't exist, the module will need to be
+        # extracted to a subdirectory.
+        repository_ctx.delete(repository_ctx.path("."))
+        output = ""
+        split = repository_ctx.attr.distribution.split("-")
+        if len(split) > 1:
+            output = split[0]
+
+        repository_ctx.download_and_extract(
+            repository_ctx.attr.urls,
+            sha256 = results.sha256,
+            stripPrefix = repository_ctx.attr.strip_prefix,
+            output = output,
+        )
+
+        includes = ["."]
+
+    repository_ctx.file("WORKSPACE.bazel", """workspace(name = "{}")""".format(
+        repository_ctx.name,
+    ))
+
+    repository_ctx.file("BUILD.bazel", _CPAN_MODULE_BUILD_FILE.format(
+        name = repository_ctx.attr.distribution,
+        hub_name = repository_ctx.attr.hub_name,
+        repo_name = "{}__{}".format(repository_ctx.attr.hub_name, repository_ctx.attr.distribution),
+        dependencies = json.encode_indent(repository_ctx.attr.dependencies, indent = " " * 4),
+        output = "{}/".format(output).lstrip("/"),
+        includes = json.encode(includes),
+    ))
+
+    return {
+        "dependencies": repository_ctx.attr.dependencies,
+        "distribution": repository_ctx.attr.distribution,
+        "hub_name": repository_ctx.attr.hub_name,
+        "name": repository_ctx.name,
+        "sha256": results.sha256,
+        "strip_prefix": repository_ctx.attr.strip_prefix,
+        "urls": repository_ctx.attr.urls,
+    }
+
+perl_cpan_archive = repository_rule(
+    doc = "A repository rule for fetching Perl modules from CPAN and instantiating a target for it.",
+    implementation = _perl_cpan_archive_impl,
+    attrs = {
+        "dependencies": attr.string_list(
+            doc = "The dependencies of the current module.",
+            mandatory = True,
+        ),
+        "distribution": attr.string(
+            doc = "The distribution of the module as described in the CPAN Metadata.",
+            mandatory = True,
+        ),
+        "hub_name": attr.string(
+            doc = "The name of the bzlmod hub repository.",
+            mandatory = True,
+        ),
+        "sha256": attr.string(
+            doc = "The expected SHA-256 of the file downloaded.",
+        ),
+        "strip_prefix": attr.string(
+            doc = "A directory prefix to strip from the extracted files.",
+        ),
+        "urls": attr.string_list(
+            doc = "A list of URLs to a file that will be made available to Bazel.",
```

https://bazel.build/rules/lib/builtins/repository_ctx#download_and_extract
Has a IMO slightly more helpful docstring.
Also it could be nice to accept both a string `url` or an iterable of strings `urls` as input? Maybe yagni.
```suggestion
            doc = "List of mirror URLs referencing the same file to download.",
```